### ISSUE TITLE
HHVM: Class already declared: PHPUnit_Framework_Error

### ISSUE THREAD

**Issue by splitbrain** [sebastianbergmann/phpunit#1684]
I'm trying to get started with testing with HipHopVM, however I have trouble even get it to run:

```
$ hhvm --version
HipHop VM 3.6.1 (rel)
Compiler: 1428401411_698854969
Repo schema: 479647c3ad87f2e920ac085aa1424e2b8101da45
```

All the available phpunit versions throw the same error:

```
$ php phpunit.phar --version
PHPUnit 4.6.4 by Sebastian Bergmann and contributors.
$ hhvm phpunit.phar --version
Fatal error: Class already declared: PHPUnit_Framework_Error in phar://phpunit-4.6.4.phar/phpunit/Util/../Framework/Error.php on line 22

$ php phpunit-beta.phar --version
PHPUnit beta-2015-04-17 by Sebastian Bergmann and contributors.
$ hhvm phpunit-beta.phar --version
Fatal error: Class already declared: PHPUnit_Framework_Error in phar://phpunit-beta-2015-04-17.phar/phpunit/Util/../Framework/Error.php on line 22

$ php phpunit-alpha.phar --version
PHPUnit alpha-2015-04-17 by Sebastian Bergmann and contributors.
$ hhvm phpunit-alpha.phar --version
Fatal error: Class already declared: PHPUnit_Framework_Error in phar://phpunit-alpha-2015-04-17.phar/phpunit/Util/../Framework/Error.php on line 22
```

Any idea what's wrong?

**Comment by machee** [comment]
Having the same trouble with HHVM 3.6.1 and PHPUnit 4.6.1-4.6.4.  PHPUnit 4.5.1 works for me as a work around.

https://phar.phpunit.de/phpunit-4.5.1.phar

I do wonder if master/nightly HHVM works with PHPUnit 4.6.4 but haven't tried it.

**Comment by thunderer** [comment]
Got the same error while running HHVM build in JoliCI:

```
Step 8 : CMD /bin/bash -c -l "cd $WORKDIR && phpunit"
 ---> Running in bd1c9df892ae
 ---> 4a4d3958f69f
Removing intermediate container bd1c9df892ae
Successfully built 4a4d3958f69f

Fatal error: Class already declared: PHPUnit_Framework_Error in phar://phpunit-4.6.4.phar/phpunit/Util/../Framework/Error.php on line 22

Fatal error: Class already declared: PHPUnit_Framework_Error in phar://phpunit-4.6.4.phar/phpunit/Util/../Framework/Error.php on line 22
```

**Comment by SimplyCorey** [comment]
Same issue here.

    Fatal error: Class already declared: PHPUnit_Framework_Error in phar://phpunit-     4.6.4.phar/phpunit/Util/../Framework/Error.php on line 22
    The command "phpunit --bootstrap tests/bootstrap.php --configuration phpunit.xml tests" exited with 255.
    Done. Your build exited with 1.

**Comment by koriym** [comment]
Same issue here, too.
```
$ phpunit --coverage-text --coverage-clover=coverage.clover
Fatal error: Class already declared: PHPUnit_Framework_Error in phar://phpunit-4.6.4.phar/phpunit/Util/../Framework/Error.php on line 22
The command "phpunit --coverage-text --coverage-clover=coverage.clover" exited with 255.
```

**Comment by kinosang** [comment]
Get the same error with HHVM 3.6.1 on Travis CI
```
$ phpunit
Fatal error: Class already declared: PHPUnit_Framework_Error in phar://phpunit-4.6.4.phar/phpunit/Util/../Framework/Error.php on line 22
The command "phpunit" exited with 255.
```

**Comment by sebastianbergmann** [comment]
The PHAR works fine with PHP. If it does not work with HHVM then that would be an HHVM issue and not a PHPUnit issue.

**Comment by kinosang** [comment]
@sebastianbergmann It works well for me with HHVM on Travis CI yesterday. I don't know if it caused by the changes of my project or others.

**Comment by zachborboa** [comment]
+label: 2015-04 build update issue?

**Comment by spiritix** [comment]
This was working yesterday, seems to be related to the last Travis CI rollout.
https://github.com/travis-ci/travis-ci/issues/3790

**Comment by philsturgeon** [comment]
I've implemented [this workaround](https://github.com/thephpleague/commonmark/commit/a0ef290ed4b4f95e7fa1a0bbc6af79c6ec6b5b1a) from @colinodell, and it just uses composer if its HHVM.

I think HHVM need to fix this, or Travis, but whatever, this got me green again.

**Comment by fredemmott** [comment]
It looks like this is triggered by:
 - swapping from an autoload-based phar to a huge pile of requires at the top
 - PHPUnit_Framework_Error is now required twice, with different paths. My guess is that HHVM's failing to normalize the paths within the phar properly, and not realizing they're the same file.

**Comment by sebastianbergmann** [comment]
Thanks for looking into this, @fredemmott.

* The "huge pile of requires at the top" is due to #1599 and works fine with PHP
* The explicit loading of `PHPUnit_Framework_Error` is due to #125 (and PHP bugs https://bugs.php.net/bug.php?id=47987 and https://bugs.php.net/bug.php?id=65322)

**Comment by stof** [comment]
@sebastianbergmann the requirement added in #125 seems to rely on the include path being set. What about using absolute paths instead (``__DIR__.'/../Framework/Error/Notice.php'``). It may solve the issue with the double requirement.

**Comment by sebastianbergmann** [comment]
@stof Good point. Can you send a pull request?

**Comment by stof** [comment]
hmm, actually, it has already been changed in https://github.com/sebastianbergmann/phpunit/commit/a3dcfcb1347be04957d38bdb292b0ffb0e3c6cb6
So this looks weird

**Comment by stof** [comment]
OK, the issue is that HHVM does not resolve paths properly inside phars, so the require_once are considering that they are 2 files: https://github.com/facebook/hhvm/issues/5215. Maybe uding ``dirname()`` rather than ``..`` would solve this

**Comment by stof** [comment]
OK, here is a PR

**Comment by joni2back** [comment]
+1

**Comment by joni2back** [comment]
I just added new php file to my project (~20min ago) (https://github.com/joni2back/php-classic/commit/bd09ad207bee47e4210055cd4e6b1918e298c858) after a long time.
So, travis test fails testing **hhvm** [travis.yml](https://github.com/joni2back/php-classic/blob/master/.travis.yml)
```
Fatal error: Class already declared: PHPUnit_Framework_Error in phar://phpunit-4.6.4.phar/phpunit/Util/../Framework/Error.php on line 22
The command "phpunit --configuration phpunit.xml --coverage-text" exited with 255.

**Comment by sebastianbergmann** [comment]
@joni2back The fix will be in PHPUnit 4.6.6.

**Comment by fredemmott** [comment]
@sebastianbergmann : yep - the pile of requires /should/ be a perfectly fine approach, and that not working is an HHVM bug. It looks like we include phar:// like any other stream, path normalization on streams is generally not a good idea, but we might need to special-case this :(
